### PR TITLE
remove hardcoded relation names

### PR DESCRIPTION
**Pull request by Jacob-Eliat-Eliat**
This is my attempt to use a constant name for each relations.

There are a could of quirks.. 

One relation doesn't have a proper class instead is a Builder, but you can feasably just give a companion object and a name to the builder.

In two different places we were doing a switch over all relations. I changed one to make everything go through the pattern matching instead of having an if-else + matching for the rest

You cannot properly enforce that a class has a companion object with a certain attribute in scala using types unfortunately. I did create a trait for it, but it doesn't really provide anything.

"shortname" already existed but I guess you'd need an instance for that to work.

### CODE REVIEW

File: src/main/scala/cognite/spark/v1/ThreeDModelsRelation.scala [hunk 1]
```diff
@@ -11,7 +11,7 @@ import org.apache.spark.sql.{Row, SQLContext}
 final case class ModelItem(id: Long, name: String, createdTime: Long)
 
 class ThreeDModelsRelation(config: RelationConfig)(val sqlContext: SQLContext)
-    extends SdkV1Relation[ThreeDModel, Long](config, "threeDModels.read") {
```
Comment (Jacob-Eliat-Eliat) [inline src/main/scala/cognite/spark/v1/ThreeDModelsRelation.scala:14]: this shortName was different for once. I think this is only used for metric so the change should have no impact.

File: src/main/scala/cognite/spark/v1/ThreeDModelRevisionsRelation.scala [hunk 2]
```diff
@@ -36,3 +36,7 @@ class ThreeDModelRevisionsRelation(config: RelationConfig, modelId: Long)(val sq
       client: GenericClient[IO]): Seq[Stream[IO, ThreeDRevision]] =
     Seq(client.threeDRevisions(modelId).list(config.limitPerPartition))
 }
+
+object ThreeDModelRevisionsRelation extends NamedRelation {
+  override val name: String = "3dmodelrevisions"
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/ThreeDModelRevisionsRelation.scala:41]: value is changing here, but looks like nothing depends on it in transformations atm :thinking:

Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/ThreeDModelRevisionsRelation.scala:41]: Maybe support wasn't added to transformations yet https://docs.cognite.com/cdf/integration/guides/transformation/write_sql_queries/ - not listed here

File: src/main/scala/cognite/spark/v1/FileContentRelation.scala [hunk 3]
```diff
@@ -144,3 +144,7 @@ class FileContentRelation(config: RelationConfig, fileExternalId: String, inferS
     }
 
 }
+
+object FileContentRelation extends NamedRelation {
+  override val name: String = "filecontent"
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/FileContentRelation.scala:149]: looks like we have a todo for file content - add datasource metrics like others, make it extend CdfRelation

File: src/main/scala/cognite/spark/v1/ThreeDModelRevisionsRelation.scala (hunk 2 above)
Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/ThreeDModelRevisionsRelation.scala:41]: does match the datasource `case` string, so better to stick to that, only shortname for metrics changes here like in the other 3d file :+1:

File: src/main/scala/cognite/spark/v1/PushdownUtilities.scala [hunk 4]
```diff
@@ -379,6 +379,10 @@ object PushdownUtilities {
   }
 }
 
+trait NamedRelation {
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/PushdownUtilities.scala:382]: nit: looks unrelated to pushdown. Also, maybe make CdfRelation extend NamedRelation, they can use the same shortname and it makes it more likely that named part is applied to new source types

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 5]
```diff
@@ -188,104 +188,108 @@ class DefaultSource
...
+      case relation: AssetHierarchyBuilder =>
+        config.onConflict match {
           case OnConflictOption.Delete =>
-            relation.delete
+            relation.delete(data)
+          case _ =>
+            relation.buildFromDf(data)
         }
+        relation
+      case relation: CdfRelation with WritableRelation =>
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: It looks to be on same level as cases by relation types so it will not be applied if any of previous ones match
May need to open in IDE old&new to double-check though

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 6]
```diff
@@ -188,104 +188,108 @@ class DefaultSource
...
 
-        val operation: Seq[Row] => IO[Unit] = config.onConflict match {
-          case OnConflictOption.Abort =>
-            relation.insert
-          case OnConflictOption.Upsert =>
-            relation.upsert
-          case OnConflictOption.Update =>
-            relation.update
+    relation match {
+      case relation: DataPointsRelationV1[_] =>
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:223]: nit: while at prettifying maybe assets&datapoints could also be extracted into functions like createSequenceRows and createFlexibleDataModelRelation :thinking:

File: src/main/scala/cognite/spark/v1/PushdownUtilities.scala (hunk 4 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/PushdownUtilities.scala:382]: This trait is for Objects so it doesn't really work, AFAIK you can't force classes to have companion objects with a variable

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 5 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: Think you are misreading this? there is two pattern matching statements now.

First ones does the case by case for each source type and create the relation

Second one does some handling depending on the type of relation, there is special handling for both Datapoints and ... I forgot which one and all the rest go through this case here

File: src/main/scala/cognite/spark/v1/PushdownUtilities.scala (hunk 4 above)
Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/PushdownUtilities.scala:382]: right, makes sense

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 5 above)
Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: could be, in web UI it is a bit confusing to see the nesting, will try in a proper UI

File: src/main/scala/cognite/spark/v1/FileContentRelation.scala (hunk 3 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/FileContentRelation.scala:149]: Ah yes, this is the companion object but you are right here :s

File: src/main/scala/cognite/spark/v1/ThreeDModelRevisionsRelation.scala (hunk 2 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/ThreeDModelRevisionsRelation.scala:41]: Ah yes, I mentionned the other one but forgot this one changes too. It does look right though!

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 6 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:223]: There is definitely more to do here, but let's keep the scope limited

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 5 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: Definitely :)

File: src/main/scala/cognite/spark/v1/FileContentRelation.scala (hunk 3 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/FileContentRelation.scala:149]: Will do in another PR if that's ok

File: src/main/scala/cognite/spark/v1/PushdownUtilities.scala (hunk 4 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/PushdownUtilities.scala:382]: Ok, I did the latter, now both the companion object and the relation itself extend namedRelation, I'm not sure if I like it better or not but it can't hurt

File: src/main/scala/cognite/spark/v1/FileContentRelation.scala (hunk 3 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/FileContentRelation.scala:146]: Will do in this PR, sounds like it's easy.

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 5 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: Ah you mean it won't be applied for Datapoints and AssetHierarchyBuilder? That is true, but it's the same as before, as it was before nested in the "else" statement

Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: Maybe it should be applied in these cases too, but that would be a behavior changes and should be looked at separately.

Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: yep, for this PR best to keep existing behavior

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 7]
```diff
@@ -188,104 +188,108 @@ class DefaultSource
...
+            .parEvalMapUnordered(maxParallelism) { chunk =>
+              operation(chunk.toVector)
+            }
+            .compile
+            .drain
+            .unsafeRunSync()
+        })
+        relation
+      case _ =>
+        sys.error(s"Resource type $resourceType does not support save")
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:292]:
```suggestion
        sys.error(s"Resource type $resourceType does not support save()")
```
maybe keep existing message

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 5 above)
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: hm, a little confusing naming from old code DefaultSource.createRelation is actually createRelationForWriting or something like it. But indeed it was checking against a list of relations that are supposed to support writing, now we check it via a trait so overall looks ok. By a quick look there were no writable sources not covered before so logic should be same as before :+1:

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 8]
```diff
@@ -188,104 +188,108 @@ class DefaultSource
...
+            import CdpConnector.ioRuntime
+            val batches = rows.grouped(batchSize).toVector
+            batches.parTraverse_(relation.delete).unsafeRunSync()
+          })
+        } else {
+          // datapoints need special handling of dataframes and batches
+          relation.insert(data, overwrite = true)
+        }
+        relation
+      case relation: AssetHierarchyBuilder =>
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:237]:
```suggestion
      case relation: AssetHierarchyBuilder with WritableRelation =>
```
looks like in spirit it should also be checked (is it valid Scala syntax?)

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 9]
```diff
@@ -188,104 +188,108 @@ class DefaultSource
...
+      case StringDataPointsRelation.name =>
+        new StringDataPointsRelationV1(config)(sqlContext)
+      case EventsRelation.name =>
+        new EventsRelation(config)(sqlContext)
+      case TimeSeriesRelation.name =>
+        new TimeSeriesRelation(config)(sqlContext)
+      case AssetsRelation.name =>
+        new AssetsRelation(config)(sqlContext)
+      case FilesRelation.name =>
+        new FilesRelation(config)(sqlContext)
```
Comment (dmivankov) [inline src/main/scala/cognite/spark/v1/DefaultSource.scala:206]: I guess that we handle it here and don't bail with "doesn't support save()" might be ok in presence of another check later on, but looks a bit odd

Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:206]: nevermind, that's files metadata, not contents

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 5 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:245]: This is also necessary to be able to call some stuff below (I believe the schema stuff notably)

Used to be this was in the "else" and then the typing of relation was enforced by the pattern matching, but in this case, we are pattern matching again on relation, which can be a AssetHierarchyBuilder which IIRC is not marked as writable, or something like that.

Basically some typing issue force this to be here, but it also makes perfect sense with the below code.

**Review by dmivankov — APPROVED** [review]
Looks like it should preserve existing behavior :+1:

File: src/main/scala/cognite/spark/v1/DefaultSource.scala (hunk 8 above)
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:237]: It doesn't implement it, but maybe it should ^^

Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:237]: Hmm, probably it should, I'll add it I guess

Comment (dmivankov) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:237]: oh, that's annoying, maybe we need a more fine-grained trait then to make it super clean

Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:237]: ah it's even worse.

It implements insert/delete/update but the signature is different (doesn't take row, takes vector of AssetUpdateSchema/AssetInsertSchema etc...) room for improvement here. I was looking at fixing this but looks too annoying to be in the same PR at elast

File: src/main/scala/cognite/spark/v1/DefaultSource.scala [hunk 10]
```diff
@@ -188,104 +188,108 @@ class DefaultSource
...
+        new LabelsRelation(config)(sqlContext)
+      case SequenceRowsRelation.name =>
+        createSequenceRows(parameters, config, sqlContext)
+      case RelationshipsRelation.name =>
+        new RelationshipsRelation(config)(sqlContext)
+      case DataSetsRelation.name =>
+        new DataSetsRelation(config)(sqlContext)
+      case FlexibleDataModelBaseRelation.name =>
+        createFlexibleDataModelRelation(parameters, config, sqlContext)
+      case _ => sys.error(s"Resource type $resourceType does not support save()")
```
[the remark this answers is not in the document]
Comment (Jacob-Eliat-Eliat) [reply src/main/scala/cognite/spark/v1/DefaultSource.scala:219]: Agreed here, but it's due to the weird traits implementation of AssetHierarchyBuilder, or lack thereof, need separate work, and it's probably a pain, there must be a reason it's done this way.
